test/testutilsversion.cpp

```diff
@@ -44,8 +44,9 @@ private slots:
     {
         using TwoDigits = Utils::Version<unsigned char, 2, 1>;
         TwoDigits();
-        TwoDigits(0);
-        TwoDigits(0, 1);
+        QCOMPARE(TwoDigits(u"0"_qs), TwoDigits(0));
+        QCOMPARE(TwoDigits(u"0.1"_qs), TwoDigits(0, 1));
+        QCOMPARE(TwoDigits(u"-1.-2"_qs), TwoDigits(-1, -2));
```

Comment:
It doesn't seem to be correct for `unsigned char`.

Reply:
Moreover, is there any **practical** reason to support negative numbers as values of version components?

Reply:
>It doesn't seem to be correct for unsigned char.

Will remove.

>Moreover, is there any practical reason to support negative numbers as values of version components?

I don't know.

Reply:
> > It doesn't seem to be correct for unsigned char.
> 
> Will remove.

In case you didn't understand me correctly:
1. `Version<unsigned char, 2>(u"-1.-2"_qs)` should throw
2. `Version<unsigned char, 2>(-1, -2)` shouldn't compile

Reply:
>In case you didn't understand me correctly:

What do you propose to do that?

Reply:
> What do you propose to do that?

Well, at least delete this test, as you suggested above, so as not to create a false impression that "negative numbers for unsigned type" is a valid case.
Ideally, it would be good to fix the constructor and parser to prevent such cases as I suggested above. But of course you don't have to do it if you don't want to (well, or at least you could add `FIXME` comments to the code that they should be fixed).

Reply:
>Ideally, it would be good to fix the constructor and parser to prevent such cases as I suggested above. But of course you don't have to do it if you don't want to (well, or at least you could add FIXME comments to the code that they should be fixed).

I'll add some comments about it to satisfy you.
Also your suggestion is not achievable with the way c++ templates are designed. Even if it is implementable the code will be monstrous.

Personally, I will remove the type variable `T` and let it just be `int`. Use `-1` as the default invalid value (treat all numbers < 0 as invalid). This way it would simplify a lot of problems and concerns.

Reply:
> Personally, I will remove the type variable `T` and let it just be `int`. Use `-1` as the default invalid value (treat all numbers < 0 as invalid in this class). This way it would simplify a lot of problems and concerns.

👍